### PR TITLE
Sql autocomplete options

### LINKED ISSUE #24452: Improve the autocomplete query
**Describe the bug**
The [autocomplete query](https://github.com/metabase/metabase/blob/2f216e72f6e335d3048c5a93bfa8b5bd8d389e86/src/metabase/api/database.clj#L415) can cause performance issues on databases with massive amount of fields, so we should take some measures for improving it

**Logs**
See ticket 11024 on our ticket system

**To Reproduce**
Can't easily reproduce unless we build some massive tables

**Expected behavior**
Don't hammer the DB on each keystroke, or enhance the query by either:
- dropping the order by
- not using a %%, between the keyword, and just use a % after the keyword
- put some delay between the keyword and the query, as users will type several letters before searching for the entity they want

**Screenshots**
NA

**Information about your Metabase Installation:**
- Your databases: ?
- Metabase version: 1.43.4, but this is historic
- Metabase internal database: Postgres 11

**Severity**
P2

**Additional context**
Workaround: block the autocomplete API, but you lose the autocomplete functionality

### PR DESCRIPTION
**Pull request by dpsutton**
Fixes #24452 

Additional context from slack:
> Autosuggestion query lockup AppDB. Regression since 43, which is causing major problems for a customer's setup. Have not been able to reproduce. Customer is on 43.4.

### Context

The sql query editor autocompletes as you are typing with a mix of fields and tables from the backend, and semantic keywords for sql presumably in the frontend. Version 43 used a "prefix" style search ( `where field.name like "input%"`). Version 44 adds a substring type search ( `where field.name like "%input%"`) and this seems to have caused a massive performance degradation as it searches the content of all strings and not just ones which match the beginning of the input. I think the instance the reported the problem had over 100,000 fields in the database.

This new behavior was added in https://github.com/metabase/metabase/pull/21887 , allowing for specifying the search type in the url

- `api/database/1/autocomplete_suggestions?search=b`
- `api/database/1/autocomplete_suggestions?prefix=b`

But the sql editor always used search.

### Solution:

Adds a defsetting `sql-editor-autocomplete-match-style` whose value can be `substring`, `prefix`, or `off`. Allows for retention of the new 44 behavior, reverting back to 43 behavior, or disabling the feature if performance remains degraded.

https://github.com/metabase/metabase/pull/21887#issuecomment-1106602270 anticipated some follow-up work for heuristics of when to switch between the two. But unfortunately issuing these queries right now is causing app-db lockup and preventing people from upgrading. So while those heuristics are formalized adding a setting to control it unequivocally seems the best, simplest, and quickest approach.

Some other contemplated fixes were to increase the timeout before issuing the query, and requiring a larger substring before firing the query. These sound like good ideas in general but still leave us open to the possibillity of a db lockup requiring a restart (full table scans for substring matches regardless of size of input)

##### Substring (default)
image

##### prefix (behavior of 43)
image
(NOTE no longer matches suBtotal)

##### off
image
(NOTE still autocompletes sql syntax but nothing related to tables/fields)

### CODE REVIEW

File: src/metabase/api/database.clj [hunk 1]
```diff
@@ -445,21 +446,44 @@
         fields (readable-fields-only (autocomplete-fields db-id match-string limit))]
     (autocomplete-results tables fields limit)))
 
+(def ^:private autocomplete-matching-options
+  "Valid options for the autocomplete types. Can match on a substring (\"%input%\"), on a prefix (\"input%\"), or reject
+  autocompletions. Large instances with lots of fields might want to use prefix matching or turn off the feature if it
+  causes too many problems."
+  #{:substring :prefix :off})
+
+(defsetting sql-editor-autocomplete-match-style
```
Comment (qnkhuat) [inline src/metabase/api/database.clj:455]:
```suggestion
(defsetting native-query-autocomplete-match-style
```
```suggestion
    (str "Matching style for native query's autocomplete. Can be \"substring\", \"prefix\", or \"off\". "
```

It works for mongo too, so I think sql editor might be misleading.

Also "native query" match with our naming for the main entities (question, native query, dashboard, collection), which I think makes it easier for users to know what this is about.

File: src/metabase/api/database.clj [hunk 2]
```diff
@@ -445,21 +446,44 @@
...
+  (deferred-tru
+    (str "Matching style for sql editor's autocomplete. Can be \"substring\", \"prefix\", or \"off\". "
```
Comment (qnkhuat) [inline src/metabase/api/database.clj:457]:
```suggestion
    (str "Matching style for native query's autocomplete. Can be \"substring\", \"prefix\", or \"off\". "
```

**Review by qnkhuat — APPROVED** [review]
It works as expected. I have one comment about the naming.

File: src/metabase/api/database.clj (hunk 1 above)
Comment (dpsutton) [reply src/metabase/api/database.clj]: good call. We've updated

**Review by escherize — APPROVED** [review]

**Pull request merged**